### DISCUSSION TITLE
Use several XML namespaces inside a .pdsc file

### DISCUSSION THREAD

**Discussion by kkistm** [Open-CMSIS-Pack/Open-CMSIS-Pack-Spec#195] [Q&A]
According to XML standard, it is possible to use several namespaces in a file with different schemas. The question is: are the standard Open CMSIS tools tolerant to several namespaces in a `.pdsc` file, with https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/blob/main/schema/PACK.xsd as a default namespace? What if there is a single namespace, again using https://github.com/Open-CMSIS-Pack/Open-CMSIS-Pack-Spec/blob/main/schema/PACK.xsd, but it is not default? Will the tools handle it properly?
Thank you in advance.

**Comment by kkistm** [comment]
Following proposition from @slhultgren, I am tagging @jkrech to get some attention to the issue. @jkrech, could you please shoot a glance at the question? Thank you in advance.

**Comment by jkrech** [comment]
I am not an expert in this area but it would be helpful to have an example demonstrating this usage.
My expectation would be that other schema and namespaces are only used within 'custom' sections that are ignored by the tools.
see: 
[Microchip.SAMS70_DFP.pdsc](https://packs.download.microchip.com/Microchip.SAMS70_DFP.pdsc)

         <environment name="microchip">
            <mchp:extension xmlns:mchp="http://crownking/pack-device-microchip-extension"
                            schemaVersion="1.0">
               <mchp:prerequisite component="MPLAB X IDE" version="[5.50-"/>
               <mchp:prerequisite component="MPLAB X IPE" version="[5.10-"/>
            </mchp:extension>
         </environment>

But maybe I am not getting your question.

**Reply by kkistm** [reply]
Thank you, it almost answers the question. I am not sure I understand what are 'custom' sections, though. The question is about how a tool, which follows Open CMSIS specification, should process such XML:

```xml
<?xml version="1.0" encoding="UTF-8"?>

<package schemaVersion="1.3"
  xmlns:xs="http://www.w3.org/2001/XMLSchema-instance"
  xs:noNamespaceSchemaLocation="PACK.xsd"
  xmlns:mysch="http://my/pack-scheme-extension">

  <name>CMSIS</name>
  <description>CMSIS (Common Microcontroller Software Interface Standard)</description>
  <vendor>ARM</vendor>
  <!-- <license>license.txt</license> -->
  <url>http://www.keil.com/pack/</url>
  <mysch:timestamp>2013-02-03 04:05:06</mysch:timestamp>

  <releases>
    <mysch:contact>JohnDoe@acme.com</mysch:contact>
    <release version="5.8.0" date="2021-06-24" mysch:detailed="85y49heruhge9ir9">
      CMSIS-Core(M): 5.5.0 (see revision history for details)
      ...
    </release>
    <release version="5.7.0" date="2020-04-09" mysch:detailed="65fnwo4i30tu3h08">
      CMSIS-Build: 0.9.0 (beta)
      ...
    </release>
</package>
```

My expectations are that if a tool is not aware about `"http://my/pack-scheme-extension"` scheme, it should ignore all tags and attributes from `mysch` namespace. The question is if my expectations are aligned with how it should behave according to the specification.

A business use-case could be to put some marketing info, for example, or any other additional information, to be processed by custom vendor tools and ignored by all other CMSIS tools.

**Comment by jkrech** [comment]
@kkistm how do I validate the schema in case I don't have access to the extension schema? 
Currently we are running this as part of the pack importer.
@edriouk it is my understanding that currently to tools read all tags but only "process" known tags. Are we making a difference between tags with and without namespaces?

**Reply by edriouk** [reply]
Our XML parser by default removes namespace prefixes in attribute names. This is configurable and we can keep the prefixes.
All elements and attributes get collected. Their special "processing" is performed outside RteModel, in tools code. The model provides generic access to all child items and attributes. There is no schema check inside the parser.

**Reply by kkistm** [reply]
@jkrech if there is no access to the extension schema, then it will be impossible to validate. This is why the schemas were often published to Web to be accessed freely.
Speaking about usage, it looks reasonable to validate only standard `PACK.xsd` elements, if it is supported by the used parser. The other elements will be validated by tools which understand a particular schema. IMHO.
